pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,971 @@
+//!
+//! If a trade fails due to specific errors whitelisted in the pallet config,
+//! then retry happens up to the maximum number of retries specified also as config.
+//! Once the max number of retries reached, the order is terminated permanently.
+//!
+//! If a trade fails due to other kind of errors, the order is terminated permanently without any retry logic.
+//!
+//! Orders are executed on block initialize and they are sorted based on randomness derived from relay chain block hash.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use cumulus_primitives_core::relay_chain::Hash;
+use frame_support::traits::DefensiveOption;
+use frame_support::{
+	ensure,
+	pallet_prelude::*,
+	traits::{Get, Len},
+	transactional,
+	weights::WeightToFee as FrameSupportWeight,
+};
+use frame_system::{ensure_signed, pallet_prelude::OriginFor, Origin};
+use hydradx_traits::pools::SpotPriceProvider;
+use hydradx_traits::{OraclePeriod, PriceOracle};
+use orml_traits::arithmetic::CheckedAdd;
+use orml_traits::MultiCurrency;
+use orml_traits::NamedMultiReservableCurrency;
+use pallet_route_executor::Trade;
+use pallet_route_executor::TradeAmountsCalculator;
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+use sp_runtime::traits::CheckedMul;
+use sp_runtime::traits::One;
+use sp_runtime::{
+	traits::{BlockNumberProvider, Saturating},
+	ArithmeticError, BoundedVec, DispatchError, FixedPointNumber, FixedU128, Permill,
+};
+use sp_std::vec::Vec;
+use sp_std::{
+	cmp::{max, min},
+	vec,
+};
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarks;
+
+pub mod types;
+pub mod weights;
+
+use weights::WeightInfo;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+use crate::types::*;
+
+type BlockNumberFor<T> = <T as frame_system::Config>::BlockNumber;
+
+pub const SHORT_ORACLE_BLOCK_PERIOD: u32 = 10;
+pub const RETRY_TO_SEARCH_FOR_FREE_BLOCK: u32 = 5;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use frame_support::weights::WeightToFee;
+
+	use frame_system::pallet_prelude::OriginFor;
+	use hydra_dx_math::ema::EmaPrice;
+	use hydradx_traits::pools::SpotPriceProvider;
+	use hydradx_traits::PriceOracle;
+	use orml_traits::NamedMultiReservableCurrency;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T>
+	where
+		<T as pallet_route_executor::Config>::Balance: From<Balance>,
+		Balance: From<<T as pallet_route_executor::Config>::Balance>,
+	{
+		fn on_initialize(current_blocknumber: T::BlockNumber) -> Weight {
+			let mut weight = <T as pallet::Config>::WeightInfo::on_initialize_with_empty_block();
+
+			let Ok(mut random_generator) = T::RandomnessProvider::generator() else {
+				Self::deposit_event(Event::RandomnessGenerationFailed {
+					block: current_blocknumber,
+				});
+				return weight;
```

Comment:
i know that this error shouldn't really happen, but this  resolution is not acceptable, scheduled schedules has to be processed somehow, either without ranom order or rescheduled to the future block otherwise they are forever stuck and have to be terminated manually

Reply:
in this case i would always fill generator with seed 0 if the hash is not available

Reply:
All right,  I will do this then 
```
			let mut random_generator = match T::RandomnessProvider::generator() {
				Ok(randomness) => randomness,
				Err(err) => {
					Self::deposit_event(Event::RandomnessGenerationFailed {
						block: current_blocknumber,
						error: err,
					});
					rand::rngs::StdRng::seed_from_u64(0)
				}
			};
			```
So we have seed from 0u64 but the error is also logged